### PR TITLE
Add tests for PAM service module

### PR DESCRIPTION
**Pull request by vvidic**
Try to dlopen pam_glome.so and test different cases using fake pam
functions.

### CODE REVIEW

**Comment by pkern** [comment]
Hm, that's a unit test. Would there have been a way to actually test the PAM dialog instead? I'm worried about the inherent assumptions we have around how the module is being called. Actually performing PAM would show that the module is usable in a normal PAM setting.

**Comment by vvidic** [comment]
On Wed, Jan 27, 2021 at 04:33:09AM -0800, Philipp Kern wrote:
> Hm, that's a unit test. Would there have been a way to actually test
> the PAM dialog instead? I'm worried about the inherent assumptions we
> have around how the module is being called. Actually performing PAM
> would show that the module is usable in a normal PAM setting.

Ah yes, autopkgtest could configure some service, but that requires a
Debian package. I will try to see if fakepam or libpam-wrapper could
be used than...

-- 
Valentin

**Comment by vvidic** [comment]
Tests updated to use libpam-wrapper now.

**Comment by vvidic** [comment]
Python tests seem to be fine now (guess it was some pip or container glitch), so this can be reviewed again.

File: login/pam_test/test [hunk 1]
```diff
@@ -0,0 +1 @@
+auth required pam_glome.so
```
Comment (burgerdev) [inline login/pam_test/test:1]: What is this file used for? I saw that we're also generating pam config files in lines 73 and 136.

**Review by burgerdev — COMMENTED** [review]
Looks good, I commented on few minor things.

File: login/pam_test.c [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    return 1;
+  }
+
+  pam_glome = getenv("PAM_GLOME");
+  if (pam_glome == NULL) {
+    puts("PAM_GLOME not found");
+    return 1;
+  }
+
+  service_file = malloc(strlen(runtime_dir) + 1 + strlen(service) + 1);
```
[the remark this answers is not in the document]
Comment (vvidic) [reply login/pam_test.c]: Last byte should be set by sprintf, but added `calloc` also to make sure :)

File: login/pam_test.c [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+  if (service_file == NULL) {
+    puts("malloc service_file failed");
+    return 1;
+  }
+
+  sprintf(service_file, "%s/%s", runtime_dir, service);
+  f = fopen(service_file, "w");
+  free(service_file);
+  if (f == NULL) {
+    puts("fopen service_file failed");
```
[the remark this answers is not in the document]
Comment (vvidic) [reply login/pam_test.c]: Added more info in the error output.

File: login/pam_test/test (hunk 1 above)
Comment (vvidic) [reply login/pam_test/test]: It was used to create a directory used for the test. Now replaced with a `custom_target` to create the required directory.

**Review by burgerdev — APPROVED** [review]

File: login/pam_test.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (burgerdev) [reply login/pam_test.c]: Hehe, I totally missed that :)

**Pull request merged**